crates/esplora/src/blocking_ext.rs

```diff
@@ -66,60 +62,75 @@ pub trait EsploraExt {
+                if status.in_best_chain && status.next_best.is_none() {
+                    break BlockId {
+                        height: status.height.expect("must have height"),
+                        hash,
+                    };
+                }
             };
-            let _ = update
-                .insert_checkpoint(update_block_id)
-                .expect("cannot repeat height here");
-            if update_block_id.hash == original_hash {
-                break;
-            }
-        }
-        let tip_at_start = BlockId {
-            height: self.get_height()?,
-            hash: self.get_tip_hash()?,
-        };
-        if let Err(failure) = update.insert_checkpoint(tip_at_start) {
-            match failure {
-                sparse_chain::InsertCheckpointError::HashNotMatching { .. } => {
-                    // there was a re-org before we started scanning. We haven't consumed any iterators, so calling this function recursively is safe.
-                    return EsploraExt::scan(
-                        self,
-                        local_chain,
-                        keychain_spks,
-                        txids,
-                        outpoints,
-                        stop_gap,
-                        parallel_requests,
-                    );
+
+            let mut new_blocks = core::iter::once((new_tip.height, new_tip.hash))
+                .collect::<BTreeMap<u32, BlockHash>>();
+
+            let mut agreement_cp = Option::<CheckPoint>::None;
+
+            for cp in prev_tip.iter().flat_map(CheckPoint::iter) {
+                let cp_block = cp.block_id();
+                let hash = self.get_block_hash(cp_block.height)?;
+                if hash == cp_block.hash {
+                    agreement_cp = Some(cp);
+                    break;
                 }
+                new_blocks.insert(cp_block.height, hash);
+            }
+
+            // check for tip changes
+            // retry if there are changes to the tip
+            let status = self.get_block_status(&new_tip.hash)?;
+            if !status.in_best_chain || status.next_best.is_some() {
+                continue 'retry;
             }
+
+            // `new_blocks` should only include blocks that are actually new
+            let new_blocks = match &agreement_cp {
+                Some(agreement_cp) => new_blocks.split_off(&(agreement_cp.height() + 1)),
+                None => new_blocks,
+            };
+            break 'retry (new_blocks, agreement_cp);
+        };
+
+        // construct checkpoints
+        for (&height, &hash) in new_blocks.iter() {
+            last_cp = Some(
+                CheckPoint::new_with_prev(BlockId { height, hash }, last_cp)
+                    .expect("heights should not conflict"),
+            );
         }
 
+        let tip = last_cp.expect("must have atleast one checkpoint");
+        let mut update = LocalUpdate::<K, ConfirmationTimeAnchor>::new(tip.clone());
+
```

Comment:
I have a feeling that this could be a bit simpler. The algorithm is.

1. Record tip
2. Get relevant transactions anchored at the tip
3. Check tip hasn't changed -- if not extend whichever previous tip still exists in chain with blocks that conflict with those that aren't and finally the new tip.

So I'd put most of this logic at the end. I feel like there might be a more concise way of expressing it.

Reply:
Do you mind being a bit more specific about the part you are talking about? Thanks

Reply:
The logic mentioned in point (3). Right now that happens at the record tip step rather than at the end.

Reply:
I've added a separate method for constructing the "update tip" (named `construct_update_tip`). It's probably not what you had in mind @LLFourn, but I think it's the only way of guaranteeing that we create a consistent view of checkpoints during the update.

The logic of finding the point of agreement can screw up consistency, if there is a double-reorg during the update. Please refer to the diagram.

![photo_2023-06-29_13-07-37](https://github.com/bitcoindevkit/bdk/assets/14065705/d3525700-0e18-4fb3-bf30-beaabc9aff3f)

Essentially, we need to atomically fetch the latest x blocks from electrum/esplora to guarantee consistency. x is the assume-final depth.

Reply:
Fixed and redesigned in a47837c0a80d9d860b4c5774a978205ca82d7e64